### PR TITLE
[ML] Propagate accurate deployment timeout

### LINKED ISSUE #109407: [ML] StartTrainedModelDeploymentAction does not return a response by the timeout value
### Description

Can pretty consistently repro with:
1. `eland_import_hub_model --url http://elastic:password@localhost:9200 --hub-model-id intfloat/multilingual-e5-large --task-type text_embedding --start`
2. `POST _ml/trained_models/intfloat__multilingual-e5-large/deployment/_start?deployment_id=intfloat__multilingual-e5-large33`

The timeout error message shows up if we add the `timeout=15s` option.  The call also succeeds if we use `waitFor=starting` which is a decent workaround, though I don't believe Kibana uses that from the UI.

Kibana is retrying the request after 30s, which is why we see the `Could not start model deployment because an existing deployment with the same id` error from Kibana.

```
[2024-06-05T16:06:37,730][INFO ][o.e.x.m.a.TransportStartTrainedModelDeploymentAction] [runTask-0] [intfloat__multilingual-e5-large] received deploy request for model [intfloat__multilingual-e5-large] and timeout [30s]
[2024-06-05T16:06:44,063][INFO ][o.e.x.m.i.d.DeploymentManager] [runTask-0] [intfloat__multilingual-e5-large] Starting model deployment of model [intfloat__multilingual-e5-large]
[2024-06-05T16:07:07,732][INFO ][o.e.x.m.a.TransportStartTrainedModelDeploymentAction] [runTask-0] [intfloat__multilingual-e5-large] received deploy request for model [intfloat__multilingual-e5-large] and timeout [30s]
```

I'm thinking:
1. we always want an error response to show up when the call times out
2. we want the error message to include something like "it's still running" or the current status
3. kibana may or may not use the `waitFor=starting` when deploying a model, I think for now we can try to make sure we finish by the `timeout` value or return a response before then, and Kibana will not retry

### PR DESCRIPTION
**Pull request by prwhelan**
When a user sets a request timeout, reduce the timeout for each consumption of that value so that the end-to-end time is more accurate to that value.

Fix #109407

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java [hunk 1]
```diff
@@ -340,6 +343,18 @@ public void onResponse(TrainedModelAssignment assignment) {
                 public void onFailure(Exception e) {
                     listener.onFailure(e);
                 }
+
+                @Override
+                public void onTimeout(TimeValue timeout) {
+                    onFailure(
+                        new ElasticsearchStatusException(
+                            "Timed out after [{}] waiting for model deployment to finish. "
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java:351]:
```suggestion
                            "Timed out after [{}] waiting for model deployment to start. "
```

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/time/RemainingTimeTests.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            TimeValue.timeValueSeconds(30)
+        );
+        assertThat(remainingTime.get(), Matchers.equalTo(TimeValue.timeValueSeconds(30)));
+        assertThat(remainingTime.get(), Matchers.equalTo(TimeValue.ZERO));
+    }
+
+    // always add the first value, which is read when RemainingTime.from is called, then add the test values
+    private Supplier<Instant> times(Instant... instants) {
+        var startTime = Stream.of(Instant.now());
+        return Stream.concat(startTime, Arrays.stream(instants)).iterator()::next;
```
Comment (davidkyle) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/time/RemainingTimeTests.java:38]: Ooo clever I will copy this in future

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/time/RemainingTime.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+     * Create a {@link Supplier} that returns a decreasing {@link TimeValue} on each invocation, representing the amount of time until
+     * the call times out.  The timer starts when this method is called and counts down from remainingTime to 0.
+     * currentTime should return the most up-to-date system time, for example Instant.now() or Clock.instant().
+     */
+    static RemainingTime from(Supplier<Instant> currentTime, TimeValue remainingTime) {
+        var timeout = currentTime.get().plus(remainingTime.duration(), remainingTime.timeUnit().toChronoUnit());
+        var maxRemainingTime = remainingTime.nanos();
+        return () -> {
+            var remainingNanos = ChronoUnit.NANOS.between(currentTime.get(), timeout);
+            return TimeValue.timeValueNanos(Math.max(0, Math.min(remainingNanos, maxRemainingTime)));
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/time/RemainingTime.java:27]: Is the logic with `maxRemainingTime` necessary when `maxRemainingTime == remainingNanos`?
Is it to handle the case where `currentTime` goes backwards?

Comment (prwhelan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/time/RemainingTime.java:27]: I saw one case where `currentTime` went backwards.  There's a monotonic clock in Transforms that we could move to commons to help prevent this, but for now I just set a max time in hopes that it doesn't happen that often :crossed_fingers:

**Review by davidkyle — APPROVED** [review]
LGTM

**Pull request merged**